crates/ruff_python_semantic/src/red_knot/types.rs

```diff
@@ -0,0 +1,705 @@
+        types.union_ty(self)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct UnionType {
+    // the union type includes values in any of these types
+    elements: FxIndexSet<Type>,
+}
+
+struct UnionTypeBuilder<'a> {
+    elements: FxIndexSet<Type>,
+    context: &'a TypingContext<'a>,
+}
+
+impl<'a> UnionTypeBuilder<'a> {
+    fn new(context: &'a TypingContext<'a>) -> Self {
+        Self {
+            context,
+            elements: FxIndexSet::default(),
+        }
+    }
+
+    fn variant(mut self, ty: Type) -> Self {
+        match ty {
+            Type::Union(union_id) => {
+                let union = union_id.lookup(self.context);
+                self.elements.extend(&union.elements);
+            }
+            _ => {
+                self.elements.insert(ty);
+            }
+        }
+
+        self
+    }
+
+    fn build(self) -> UnionType {
+        UnionType {
+            elements: self.elements,
+        }
+    }
+}
+
+#[newtype_index]
+pub struct LocalIntersectionTypeId;
+
+impl LocalTypeId for LocalIntersectionTypeId {
+    type Ty = IntersectionType;
+
+    fn lookup_local(self, types: &TypeInference) -> &Self::Ty {
+        types.intersection_ty(self)
+    }
+}
+
+// Negation types aren't expressible in annotations, and are most likely to arise from type
+// narrowing along with intersections (e.g. `if not isinstance(...)`), so we represent them
+// directly in intersections rather than as a separate type. This sacrifices some efficiency in the
+// case where a Not appears outside an intersection (unclear when that could even happen, but we'd
+// have to represent it as a single-element intersection if it did) in exchange for better
+// efficiency in the within-intersection case.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct IntersectionType {
+    // the intersection type includes only values in all of these types
+    positive: FxIndexSet<Type>,
+    // the intersection type does not include any value in any of these types
+    negative: FxIndexSet<Type>,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct LocalModuleTypeId;
+
+impl LocalTypeId for LocalModuleTypeId {
+    type Ty = ModuleType;
+
+    fn lookup_local(self, types: &TypeInference) -> &Self::Ty {
+        types.module_ty()
+    }
+}
+
+impl TypeId<LocalModuleTypeId> {
+    fn member(self, context: &TypingContext, name: &Name) -> Option<Type> {
+        context.public_symbol_ty(self.scope.file(context.db), name)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct ModuleType {
+    file: VfsFile,
+}
+
+/// Context in which to resolve types.
+///
+/// This abstraction is necessary to support a uniform API that can be used
+/// while in the process of building the type inference structure for a scope
+/// but also when all types should be resolved by querying the db.
+pub struct TypingContext<'a> {
+    db: &'a dyn Db,
+
+    /// Local type inference scopes that are in the process of being built.
```

Comment:
From a type type inference point. What's the type of `b.B` in `C`'s bases? Is it `Unknown` or should it be the actual type of `B`? 

What happens with a cycle today (and I should have mentioned in the summary that I didn't look at cycles) is that Salsa will panic. Salsa's solution to cycles is that a query can define a fallback method that gets called when it finds a cycle to break the cycle. I haven't looked to closely at what we're allowed to do in the cycle resolution. We're probably not allowed to call any queries (or at least, only call queries that we are 100% sure can't create cycles). 

If the type of `a.A` is `Unknown`, than I think the cycle resolution is easy. We can just return an empty `TypeInference` result and change `symbol_ty`, `expression_ty` etc. to return `Unknown` or `Unbound` if there's no entry in the list. 

If, however, we need to have a more specific type, than this could be challenging indeed. I don't think Salsa gives us a pointer of why and where the cycle occurred and it isn't always as simple as in the given example. A cycle is also possible through many transitive dependencies. 

I'm not sure if we can maintain a stack or similar inside of a Salsa query because Salsa queries are, ideally, idempotent and don't rely on outside information, or it becomes very easy to break the automatic caching. 
One possibility is to have a specific marker type for cases where we encounter cycles. A "Lazy" or similar. I would need to think more about it. But the idea would be to defer the resolution of these types (and nodes) until query time.

Reply:
There should not be any Unknown types in this example.

There is no cycle here from the point of view of Python, or from the point of view of definition granularity. We only create a cycle here by forcing scope-at-a-time resolution of types.

I think the answer we need here is not related to "how do we handle cycles in Salsa" -- I think we need to improve our granularity so that this case is not a cycle in Salsa at all.

Reply:
I spent some more time reading about Salsa's cycle recovery and yes, I think we can't depend on it. 

What I understand from the documentation is that Salsa aborts one query in the cycle and replaces it with a fallback value. In our case that would mean that an entire module doesn't get type checked.

https://salsa-rs.github.io/salsa/cycles/fallback.html